src/doc.rs

```diff
@@ -0,0 +1,354 @@
+use super::ast::{Arg, ArgNames, Statement, TypeAst, TypedModule};
+use itertools::Itertools;
+use std::collections::HashMap;
+#[derive(Debug, PartialEq)]
+struct RawComment {
+  pub line_no: usize,
+  pub content: String,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DocBlockManager {
+  pub(self) current_block: Vec<RawComment>,
+  pub(self) blocks: Vec<Vec<RawComment>>,
+  num_comments: usize,
+}
+
+impl DocBlockManager {
+  pub fn new() -> Self {
+    DocBlockManager {
+      current_block: vec![],
+      blocks: vec![],
+      num_comments: 0,
+    }
+  }
+
+  pub fn block_count(self: &Self) -> usize {
+    if self.current_block.is_empty() {
+      self.blocks.len()
+    } else {
+      self.blocks.len() + 1
+    }
+  }
+
+  pub fn comment_count(self: &Self) -> usize {
+    self.num_comments
+  }
+
+  pub(self) fn find_block_for_line(self: &Self, line_no: usize) -> Option<String> {
+    for block in &self.blocks {
+      if block
+        .last()
+        .and_then(|b: &RawComment| Some(b.line_no == line_no - (b.content.len() + 5)))
+        .unwrap_or(false)
+      {
+        return Some(block.iter().map(|b| b.content.to_string()).join("\n"));
+      }
+    }
+
+    if self
+      .current_block
+      .last()
+      .and_then(|b: &RawComment| Some(b.line_no == line_no - (b.content.len() + 5)))
+      .unwrap_or(false)
+    {
+      return Some(
+        self
+          .current_block
+          .iter()
+          .map(|b| b.content.to_string())
+          .join("\n"),
+      );
+    }
+
+    None
+  }
+
+  pub fn add_line(self: &mut Self, line_no: usize, content: String) {
+    match self.current_block.last() {
+      Some(RawComment {
+        line_no: l,
+        content: _,
+      }) => {
+        let new_line = RawComment {
+          line_no: line_no,
+          content,
+        };
+        if (line_no - 1) == *l {
+          self.current_block.push(new_line);
+        } else {
+          let copy = std::mem::replace(&mut self.current_block, vec![]);
+          self.blocks.push(copy);
+          self.current_block = vec![];
+          self.current_block.push(new_line);
+        }
+      }
+      None => {
+        self.current_block.push(RawComment { line_no, content });
+      }
+    }
+    self.num_comments = self.num_comments + 1;
+  }
+
+  pub fn gen_doc_chunk(self: &Self, module: &TypedModule) -> EEP48DocChunk {
```

Comment:
I don't think this is the right way to approach this. If we operate on the raw AST we only have basic information about function names etc as we have not done any of the static analysis yet (such as type checking).

We could instead build the documentation from the output of the `project::compile` function which returns `Result<Vec<Compiled>, Error>`. The `Compiled` struct has lots of useful information.

Reply:
I thought that the output of infer_module had the type information?

Reply:
Oh I see! We're iterating over the typed module. My bad.

I think it'd be better to use the `Compiled` struct too as it is all the static information collected into a more compact data structure

Reply:
I didn't really find that `Compiled` had that much more useful data in it, and this let me generate the docs chunks as we are going without changing the `project::compile` function's signature too much.  This prevents a separate pass from having to happen.